File: flutter_twilio_conversations/ios/flutter_twilio_conversations/Sources/flutter_twilio_conversations/Methods/ChatClientMethods.swift
import Flutter
import Foundation
import TwilioConversationsClient

public class ChatClientMethods {
    public static func updateToken(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let arguments = call.arguments as? [String: Any?] else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing parameters", details: nil))
        }

        guard let token = arguments["token"] as? String else {
            return result(FlutterError(code: "MISSING_PARAMS", message: "Missing 'token' parameter", details: nil))
        }

        let flutterResult = result
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.updateToken(token, completion: {(result: TCHResult) -> Void in
            if result.isSuccessful {
                SwiftTwilioConversationsPlugin.debug("ChatClientMethods.updateToken => onSuccess")
                flutterResult(nil)
            } else {
                // Always resolve the Dart await, even when the SDK reports failure
                // without an error object.
                let error = result.error as NSError?
                SwiftTwilioConversationsPlugin.debug("ChatClientMethods.updateToken => onError: \(String(describing: error))")
                flutterResult(FlutterError(
                    code: "\(error?.code ?? 0)",
                    message: "Failed to update token: \(error?.description ?? "unknown error")",
                    details: nil))
            }
            } as TCHCompletion)
    }

    public static func shutdown(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        tearDownClient()
        result(nil)
    }

    /// Shuts down the current client (if any) and then clears every plugin-level reference to
    /// it, so later method calls fail fast instead of dereferencing a dead client. Requesting
    /// shutdown() before dropping our references gives the SDK an orderly teardown, but the
    /// transport teardown continues asynchronously on the SDK's own queues — protection against
    /// the over-release inside shutdownTransport() (twilio/conversations-ios#54, BK-6201) comes
    /// from the SDK bump to 4.0.9, not from this ordering.
    ///
    /// Note for Dart-side callers: the plugin's Dart layer caches Channel objects statically and
    /// only clears them in ChatClient.shutdown(). Always call shutdown() before create() —
    /// create() tears down any leftover native client as a safety net, but Dart channels that
    /// survive it will have dead native stream handlers and silently stop receiving events.
    public static func tearDownClient() {
        SwiftTwilioConversationsPlugin.debug("ChatClientMethods.tearDownClient => shutting down client")
        SwiftTwilioConversationsPlugin.chatListener?.chatClient?.shutdown()
        SwiftTwilioConversationsPlugin.chatListener?.chatClient = nil
        SwiftTwilioConversationsPlugin.chatListener = nil

        for (_, eventChannel) in SwiftTwilioConversationsPlugin.channelChannels {
            eventChannel.setStreamHandler(nil)
        }
        SwiftTwilioConversationsPlugin.channelChannels.removeAll()
        SwiftTwilioConversationsPlugin.channelListeners.removeAll()
    }
}

File: flutter_twilio_conversations/android/src/main/kotlin/twilio/flutter/twilio_conversations/methods/ChatClientMethods.kt
package twilio.flutter.twilio_conversations.methods

import com.twilio.util.ErrorInfo
import com.twilio.conversations.StatusListener
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import twilio.flutter.twilio_conversations.TwilioConversationsPlugin
import android.util.Log

object ChatClientMethods {
    fun updateToken(call: MethodCall, result: MethodChannel.Result) {
        val token = call.argument<String>("token")
                ?: return result.error("ERROR", "Missing 'token'", null)

        TwilioConversationsPlugin.chatClient?.updateToken(token, object : StatusListener {
            override fun onSuccess() {
                Log.d("TwilioInfo", "ChatClientMethods.updateToken => onSuccess")
                result.success(null)
            }

            override fun onError(errorInfo: ErrorInfo) {
                Log.d("TwilioInfo", "ChatClientMethods.updateToken => onError: $errorInfo")
                result.error("${errorInfo.code}", errorInfo.message, errorInfo.status)
            }
        })
    }

    fun shutdown(pluginInstance: TwilioConversationsPlugin, call: MethodCall, result: MethodChannel.Result) {
        return try {
            tearDownClient(pluginInstance)
            result.success(null)
        } catch (err: Exception) {
            result.error("ERROR", err.message, null)
        }
    }

    // Shuts down the current client (if any) and clears every plugin-level reference to it,
    // mirroring the iOS teardown so both platforms behave the same once the app starts
    // calling ChatClient#shutdown (BK-6201). See the iOS tearDownClient doc for the
    // shutdown-before-create contract expected of Dart callers.
    fun tearDownClient(pluginInstance: TwilioConversationsPlugin) {
        Log.d("TwilioInfo", "ChatClientMethods.tearDownClient => shutting down client")
        TwilioConversationsPlugin.chatClient?.removeAllListeners()
        TwilioConversationsPlugin.chatClient?.shutdown()
        TwilioConversationsPlugin.chatClient = null
        pluginInstance.chatListener = null

        pluginInstance.channelChannels.values.forEach { it.setStreamHandler(null) }
        pluginInstance.channelChannels.clear()
        pluginInstance.channelListeners.clear()
    }
}
